src/utils/units.hpp

```diff
@@ -0,0 +1,191 @@
+  Real GetLengthCodeToPhysical() const { return length_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetLengthPhysicalToCode() const { return 1. / length_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetTimeCodeToPhysical() const { return time_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetTimePhysicalToCode() const { return 1. / time_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassCodeToPhysical() const { return mass_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassPhysicalToCode() const { return 1. / mass_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetSpeedCodeToPhysical() const { return length_ / time_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetSpeedPhysicalToCode() const { return time_ / mass_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyCodeToPhysical() const { return energy_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyPhysicalToCode() const { return 1. / energy_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetNumberDensityCodeToPhysical() const { return number_density_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetNumberDensityPhysicalToCode() const { return 1. / number_density_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyDensityCodeToPhysical() const { return energy_ * number_density_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetEnergyDensityPhysicalToCode() const { return 1. / (energy_ * number_density_); }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassDensityCodeToPhysical() const { return mass_ * number_density_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetMassDensityPhysicalToCode() const { return 1. / (mass_ * number_density_); }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetOpacityCodeToPhysical() const { return length_ * length_ / mass_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetOpacityPhysicalToCode() const { return mass_ / (length_ * length_); }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetSpecificHeatCodeToPhysical() const { return energy_ / mass_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetSpecificHeatPhysicalToCode() const { return mass_ / energy_; }
+
+ private:
+  // Unit conversion factors from code to physical units
+  // e.g. length_ has units of cm when using CGS as physical unit system
+  // Temperature is always Kelvin
+  Real length_;
+  Real time_;
+  Real mass_;
+
+  Real energy_;
+  Real number_density_;
+
+  UnitSystem unit_system_;
+};
+
+class Constants {
+ public:
+  Constants() = delete;
+
+  KOKKOS_FUNCTION
+  Constants(Units &units);
+
+  KOKKOS_FUNCTION
+  Constants(const Constants &other) {}
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetGPhysical() const { return G_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetGCode() const { return G_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetKBPhysical() const { return kb_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetKBCode() const { return kb_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetCPhysical() const { return c_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetCCode() const { return c_code_; }
+
+  KOKKOS_INLINE_FUNCTION
+  Real GetHPhysical() const { return h_; }
+  KOKKOS_INLINE_FUNCTION
+  Real GetHCode() const { return h_code_; }
```

Comment:
I don't know if every actually need `h`?

Reply:
It may show up with energy <-> frequency for photons somewhere